### PR TITLE
Docs/hdk hash module

### PR DESCRIPTION
**Pull request by jost-s**
### Summary

Explain how hashes in Holochain are composed and its components on the module page for `hdk::hash`.

### TODO:
- [x] CHANGELOG(s) updated with appropriate info
- [x] Just before pressing the merge button, ensure new entries to CHANGELOG(s) are still under the _UNRELEASED_ heading

### CODE REVIEW

File: crates/hdk/src/hash.rs [hunk 1]
```diff
@@ -1,3 +1,109 @@
...
+//! [Base64 scheme](https://developer.mozilla.org/en-US/docs/Glossary/Base64). Encoding the example public agent key in Base64 results in:
+//!
+//! ```
+//! hCAkJ9p+IlfMpeP/HeygQt2jqHDXu4+YRAQezq3L0m9nz3wCa0Mh
+//! ```
+//!
+//! ### Self-identifying hash encoding
+//!
+//! Following the [Multibase protocol](https://github.com/multiformats/multibase), hashes in Holochain self-identify its encoding in Base64.
+//! All hashes in text format are prefixed with a `u`, to identify them as [`base64url`](https://github.com/multiformats/multibase/blob/master/multibase.csv#L23).
```
Comment (neonphog) [inline crates/hdk/src/hash.rs:30]: this line is 162 characters long... our rustfmt currently uses the default max_width of 100 characters (but alas it doesn't reformat comments), I personally use an editor with a width of 80 characters.

I vote for keeping comments generally under at least 100 characters with occasional exceptions for urls, etc. And I personally try to keep them under 80.

File: crates/hdk/src/hash.rs [hunk 2]
```diff
@@ -1,3 +1,109 @@
...
+//!
+//!
+//! ### Breakdowns of example
+//!
+//! Breakdown of the example agent pub key as byte array in decimal notation:
+//!
+//! | 132 32           | 36            | 39 218 126 34 87 204 165 227 255 29 236 160 66 221 163 168 112 215 187 143 152 68 4 30 206 173 203 210 111 103 207 124 2 107 67 33        |
+//! | ---------------- | ------------- | ----------------------------------------------------------------------------------------------------------------------------------------- |
+//! | type             | length        | hash                                                                                                                                      |
+//! | public agent key | 36 bytes long | Blake2b hash, 36 bytes long                                                                                                               |
```
Comment (neonphog) [inline crates/hdk/src/hash.rs:97]: > Blake2b hash, 36 bytes long

The blake2b hash is only the first 32 bytes. The final 4 are the dht "location" bytes, which can also serve as an integrity check of the hash itself. We interpret these final 4 bytes as a `u32` to identify a particular agent's "neighborhood" in a circle defined by this wrapping u32 value, similarly, the location of an entry or header is used to determine who is responsible for the value by mapping it to the same space.

The location is calculated by taking a 128 bit (16 byte) blake2b hash of the preceding 32 bytes, then progressively xor-ing them into just 4 bytes:

```rust
/// internal compute the holo dht location u32
pub fn holo_dht_location_bytes(data: &[u8]) -> Vec<u8> {
    // Assert the data size is relatively small so we are
    // comfortable executing this synchronously / blocking tokio thread.
    assert_eq!(32, data.len(), "only 32 byte hashes supported");

    let hash = blake2b_128(data);
    let mut out = vec![hash[0], hash[1], hash[2], hash[3]];
    for i in (4..16).step_by(4) {
        out[0] ^= hash[i];
        out[1] ^= hash[i + 1];
        out[2] ^= hash[i + 2];
        out[3] ^= hash[i + 3];
    }
    out
}
```

**Review by neonphog — CHANGES_REQUESTED** [review]
requesting a change and suggesting an addition, but it's looking good :+1:

File: crates/hdk/src/hash.rs (hunk 1 above)
Comment (jost-s) [reply crates/hdk/src/hash.rs:30]: Sure, I hard wrapped the doc comments at 80 chars.

File: crates/hdk/src/hash.rs [hunk 3]
```diff
@@ -1,3 +1,157 @@
...
+//!
+//!
+//! ### Breakdowns of example
+//!
+//! Breakdown of the example agent pub key as byte array in decimal notation:
+//!
+//! | type                   | length        | hash                                                                                                                               |
+//! | ---------------------- | ------------- | ---------------------------------------------------------------------------------------------------------------------------------- |
+//! | 132 32                 | 36            | 39 218 126 34 87 204 165 227 255 29 236 160 66 221 163 168 112 215 187 143 152 68 4 30 206 173 203 210 111 103 207 124 2 107 67 33 |
+//! | public agent key       | 36 bytes long | Blake2b hash, 36 bytes long                                                                                                        |
```
Comment (neonphog) [inline crates/hdk/src/hash.rs:145]:
```suggestion
//! | type                   | length        | hash                                                                                                                   | dht location   |
//! | ---------------------- | ------------- | ---------------------------------------------------------------------------------------------------------------------- | -------------- |
//! | 132 32                 | 36            | 39 218 126 34 87 204 165 227 255 29 236 160 66 221 163 168 112 215 187 143 152 68 4 30 206 173 203 210 111 103 207 124 | 2 107 67 33    |
//! | public agent key       | 36 bytes long | Blake2b hash, 32 bytes long                                                                                            | u32: 558066434 |
```

File: crates/hdk/src/hash.rs [hunk 4]
```diff
@@ -1,3 +1,157 @@
...
+//!
+//!
+//! Breakdown of the example agent pub key encoded as Base64:
+//!
+//! | Multibase encoding   | type + length                       | hash                                                    |
+//! | -------------------- | ----------------------------------- | ------------------------------------------------------- |
+//! | u                    | hCAk                                | J9p-IlfMpeP_HeygQt2jqHDXu4-YRAQezq3L0m9nz3wCa0Mh        |
+//! | base64url no padding | public agent key of 36 bytes length | Base64 encoding of Blake2b hash                         |
```
Comment (neonphog) [inline crates/hdk/src/hash.rs:153]:
```suggestion
//! | Multibase encoding   | type + length                       | hash + dht location                                     |
//! | -------------------- | ----------------------------------- | ------------------------------------------------------- |
//! | u                    | hCAk                                | J9p-IlfMpeP_HeygQt2jqHDXu4-YRAQezq3L0m9nz3wCa0Mh        |
//! | base64url no padding | public agent key of 36 bytes length | Base64 encoding of Blake2b hash + location              |
```

File: crates/hdk/src/hash.rs (hunk 2 above)
Comment (neonphog) [reply crates/hdk/src/hash.rs]: the u32 location defines where on the circle an agent's arc is. currently, it defines the center of the arc, michael is currently working on code that will switch it to defining the left side of the arc

"neighborhood" is a very loose term, roughly meaning a group of agents that have overlapping arcs

"authority" is actually related to dht ops, so yeah, what i actually meant was if you get the location u32 from whatever hash corresponds to the dht op and it happens to fall within an agent's defined arc

**Review by neonphog — APPROVED** [review]
couple suggestions for the example figures, but with or without that, LGTM

**Comment by jost-s** [comment]
> couple suggestions for the example figures, but with or without that, LGTM

Thanks for spotting that, I had overlooked it.

**Comment by pdaoust** [comment]
sorry I'm so late to this. I think it looks awesome. I have a few small suggestions; @jost-s should I just create a new PR against `develop?

**Comment by pdaoust** [comment]
I can't for the life of me remember what suggestions I had, so I'm not gonna worry about it 😅

**Comment by jost-s** [comment]
Lolz! I take it as a good sign. All doubts vanished.

**Pull request merged**